File: src/main/java/gg/uhc/migration/configuration/RangeParser.java
package gg.uhc.migration.configuration;

import com.google.common.base.Optional;
import com.google.common.base.Preconditions;
import com.google.common.collect.BoundType;
import com.google.common.collect.Range;

public class RangeParser {

    class RangeBound {

        protected final double bound;
        protected final BoundType boundType;

        public RangeBound(double bound, BoundType type) {
            this.bound = bound;
            this.boundType = type;
        }
    }

    /**
     * Reads a range from a given string
     *
     * @param rangeString the string to read from
     * @return the parsed range
     * @throws IllegalArgumentException on invalid range string
     */
    public Range<Double> readRange(String rangeString) {
        // notation is formatted in style of [x1..x2)
        // split it into left and right bounds
        String[] bounds = rangeString.split("\\.\\.");

        // must be 2 parts
        if (bounds.length != 2) throw new IllegalArgumentException("Missing center .. in region definition");

        Optional<RangeBound> left = readLeftBound(bounds[0]);
        Optional<RangeBound> right = readRightBound(bounds[1]);

        boolean unboundedLeft = !left.isPresent();
        boolean unboundedRight = !right.isPresent();

        if (unboundedLeft && unboundedRight) {
            return Range.all();
        }

        if (unboundedLeft) {
            return Range.upTo(right.get().bound, right.get().boundType);
        }

        if (unboundedRight) {
            return Range.downTo(left.get().bound, left.get().boundType);
        }

        return Range.range(left.get().bound, left.get().boundType, right.get().bound, right.get().boundType);
    }

    protected Optional<BoundType> readLeftBoundType(char type) {
        switch (type) {
            case '[':
                return Optional.of(BoundType.CLOSED);
            case '(':
                return Optional.of(BoundType.OPEN);
            default:
                return Optional.absent();
        }
    }

    protected Optional<BoundType> readRightBoundType(char type) {
        switch (type) {
            case ']':
                return Optional.of(BoundType.CLOSED);
            case ')':
                return Optional.of(BoundType.OPEN);
            default:
                return Optional.absent();
        }
    }

    protected Optional<RangeBound> readLeftBound(String bound) {
        return readBound(bound, true);
    }

    protected Optional<RangeBound> readRightBound(String bound) {
        return readBound(bound, false);
    }

    protected Optional<RangeBound> readBound(String bound, boolean left) {
        // check for open lower bound
        if (left && bound.equalsIgnoreCase("(-Inf")) {
            return Optional.absent();
        }

        // check for open higher bound
        if (!left && bound.equalsIgnoreCase("+Inf)")) {
            return Optional.absent();
        }

        Preconditions.checkArgument(bound.length() > 1, "Must be at least 2 characters long");

        // grab the bound type character
        Optional<BoundType> type = left ? readLeftBoundType(bound.charAt(0)) : readRightBoundType(bound.charAt(bound.length() - 1));
        Preconditions.checkArgument(type.isPresent(), "Invalid bound range type character");

        double range;
        try {
            range = Double.parseDouble(left ? bound.substring(1) : bound.substring(0, bound.length() - 1));
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Invalid range number", e);
        }

        return Optional.of(new RangeBound(range, type.get()));
    }
}

File: src/main/java/gg/uhc/migration/configuration/RangeParser.kt
package gg.uhc.migration.configuration

import com.google.common.base.Preconditions.checkArgument
import com.google.common.collect.BoundType
import com.google.common.collect.Range

private data class RangeBound(val bound: Double, val boundType: BoundType)

private fun Char.readLeftBoundType() = when (this) {
    '[' -> BoundType.CLOSED
    '(' -> BoundType.OPEN
    else -> throw IllegalArgumentException("Invalid bound range type character, expected either [ or (")
}

private fun Char.readRightBoundType() = when (this) {
    ']' -> BoundType.CLOSED
    ')' -> BoundType.OPEN
    else -> throw IllegalArgumentException("Invalid bound range type character, expected either ] or )")
}

private fun String.readBound(left: Boolean): RangeBound? {
    checkArgument(this.length > 1, "Must be at least 2 characters long")

    // check for open lower bound
    if (left && this.equals("(-Inf", ignoreCase = true))
        return null

    // check for open higher bound
    if (!left && this.equals("+Inf)", ignoreCase = true))
        return null

    // grab the bound type character and text number
    val type: BoundType
    val textRange: String
    if (left) {
        type = this.first().readLeftBoundType()
        textRange = this.drop(1)
    } else {
        type = this.last().readRightBoundType()
        textRange = this.dropLast(1)
    }

    val range: Double = try {
        textRange.toDouble()
    } catch (e: NumberFormatException) {
        throw IllegalArgumentException("Invalid range number: $textRange", e)
    }

    return RangeBound(range, type)
}

private fun String.readLeftBound(): RangeBound? = this.readBound(true)
private fun String.readRightBound(): RangeBound? = this.readBound(false)

fun String.readRange(): Range<Double> {
    // notation is formatted in style of [x1..x2)
    // split it into left and right bounds
    val bounds = this
        .split("..")
        .filterNot(String::isEmpty)

    // must be 2 parts
    if (bounds.size != 2)
        throw IllegalArgumentException("Missing single `..` from region definition")

    val left = bounds[0].readLeftBound()
    val right = bounds[1].readRightBound()

    return when {
        left == null && right == null -> Range.all<Double>()
        left == null -> Range.upTo(right!!.bound, right.boundType)
        right == null -> Range.downTo(left.bound, left.boundType)
        else -> Range.range(left.bound, left.boundType, right.bound, right.boundType)
    }
}